### DISCUSSION TITLE
Several questions

### DISCUSSION THREAD

**Discussion by lowlighter** [c4spar/cliffy#225] [Q&A] [answered]
Hello, thanks a lot for this awesome library !

I've been experimenting a bit with cliffy and I have several questions that I couldn't find answer in documentation.

**1) Is there a way do to command aliases ?**

Something like:
```ts
new Command()
  .command("docs, documentation", new Command())

//Current workaround
const docs = new Command()
new Command()
  .command("docs", docs)
  .command("documentation", documentation)
```

**2) Is it possible to handle unknown options ?**

For instance, if I was passing `--option.test.a` and `--option.test.b`, I would get the following:
```
  error: Unknown option "--option.test.a". Did you mean option "--help"?
```

But would it be possible to catch all `--option.test.*` options and use a custom handlers and get `{options:{test:{a, b}}` for example ?

**3) How to correctly handle "empty commands"**

Say I have the following:
```ts
await new Command()
  .command("sub", new Command()
    .action(() => console.log("hello"))
  )
  .parse(Deno.args)
```

How do I make it fall it errors if "sub" command is not provided ? Currently it justs ouputs nothing. I'd like to have a similiar output as when passing an unknown command like:
```
  error: Unknown command "test". Did you mean command "sub"?
```

Edit: Found out that adding a `.arguments("<command>")` do exactly that, though it does not handle "invalid command" (it outputs nothing)

**4) Is it possible to make options conflictings with arguments**

Maybe this is out of scope, but I was wondering whether it's possible to make options conflicts with some command optional arguments, like:
```ts
await new Command()
  .command("sub <arg1> [arg2]", new Command()
    .option("--ok-but-only-if-no-arg2", {conflicts:["arg2"]))
  )
  .parse(Deno.args)
``` 

**5) Is it by choice that GitHub provider is not exported and must be imported from file instead ?**
https://github.com/c4spar/deno-cliffy/blob/d93ad7d8033dd5a630e6bad05722f96392e7c54e/command/upgrade/mod.ts#L1-L4

**6) Is there some way to remove "v" in version name ?**

Setting version to `.version("0.1.0")` will automatically prepend a "v", which seems redundant in help since it's already prefixed with a "Version" word
```
  Version: v0.1.0
```

Thanks for all answers!

**Comment by c4spar** [comment, marked as answer]
Hi @lowlighter, i'm glad you like cliffy!

**1) Is there a way do to command aliases ?**
Aliases are already supported. But it looks like they are not well documented. Will update the documentation. There are two ways to specify command aliases:

```ts
cmd.command("rm, remove <dir>")
# or
cmd.command("rm <dir>").alias("remove")
```

**2) Is it possible to handle unknown options ?**
Unknown options are currently not supported but would make sence! Can you open an issue for this? Than i will check how we can implement something like this.
Maybe it would also be nice to specify a regex like pattern for option/command names. But we have to ensure that this doesn't has a huge impact to the perfomance. But somethink like `--option.test.*` should be doable. And mybe somthink like `.allowUnkownOptions()` to suppress errors for unknown options and parse all unknown options as string.

**3) How to correctly handle "empty commands"**
This should already be the case, if not it's a bug. But for me it seems to be working fine. If i run your example with `COMMAND sub` it outputs `hello`. If i run your example with `COMMAND foo` it outputs `error: Unknown command "foo". Did you mean command "sub"?`.
But if you add `.arguments("<command>")` than it outputs nothing but this is working as designed, because this meens, you allow to pass any arguments to this command and you have to handle arguments by your self. If you don't specify `.arguments()` than an error is thrown if the argument doesn't match any sub command.
For example:
```ts
await new Command()
    .arguments("<some-other-argument>")
    .action((_, arg) => console.log(`hello from main command with argument: ${arg}`))
    .command("sub", "some sub command")
    .action(() => console.log("hello from sub command"))
    .parse();
```

**4) Is it possible to make options conflictings with arguments**
This is cerrently not supported but i think it's worth to open a new issue for this!

**5) Is it by choice that GitHub provider is not exported and must be imported from file instead ?**
No, it should be exported the same way as all other providers. I think i just forgot to export it.

**6) Is there some way to remove "v" in version name ?**
Not yet, but yeah it makes sence to remove the v previfx from the help text or maybe better just show the exact same version string that was provided with the `.version()` method because this is mostly the name of the git tag. good catch!

Thx for all the points @lowlighter 👍 Can you create separate issues for all the bugs and feature request? Than we can better discuss it and start working on them. Otherwise i can also do it.

**Reply by lowlighter** [reply]
1. Nice! Indeed <kbd>Ctrl + f</kbd> in documentation for `alias` yield nothing and I didn't think of trying the same syntax as option aliases but that makes a lot of sense!
2. Created #227
3. Ok, I was placing `action()` in last, as I was using the notation where you pass an instance of `Command` instead of chaining `action()` so I didn't thought that order mattered:
```js
await new Command()
  .arguments("<action>")
  .command("sub1", new Command().action(() => console.log("hello")))
  .command("sub2", new Command().action(() => console.log("world")))
  .action(() => { throw new Error() })
  .parse()
```
But indeed calling `action()` just after `arguments()` solves it, thanks for clarifying!

4. Created #228
5. Created #229
6. Created #230

Thanks a lot for your answers and your help!
Great work 👍 !

**Reply by c4spar** [reply]
@lowlighter thx a lot for creating the issues 👍

3. Yep the order is important. In your example the last action handle will override the action handler from the `sub2` sub-command.
```ts
await new Command()
  .arguments("<action>")
  .command("sub1", new Command().action(() => console.log("hello")))
  .command("sub2", new Command().action(() => console.log("world")))
  .action(() => { throw new Error() }) // <-- This overrides the action handler from "sub2"
  .parse()
```
All methods that act on a command, such as name, description, action, etc., ... always act on the last created command or subcommand. This is described here: https://github.com/c4spar/deno-cliffy/tree/v0.19.2/command#-commands
